x/supply/client/rest/query.go

```diff
@@ -0,0 +1,88 @@
+package rest
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/mux"
+
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/types/rest"
+	"github.com/cosmos/cosmos-sdk/x/supply/types"
+)
+
+// RegisterRoutes registers staking-related REST handlers to a router
+func RegisterRoutes(cliCtx context.CLIContext, r *mux.Router) {
+	registerQueryRoutes(cliCtx, r)
+}
+
+func registerQueryRoutes(cliCtx context.CLIContext, r *mux.Router) {
+	// Query the total supply of coins
+	r.HandleFunc(
+		"/supply/total/",
+		totalSupplyHandlerFn(cliCtx),
+	).Methods("GET")
+
+	// Query the supply of a single denom
+	r.HandleFunc(
+		"/supply/total/{denom}",
+		supplyOfHandlerFn(cliCtx),
+	).Methods("GET")
+}
+
+// HTTP request handler to query the total supply of coins
+func totalSupplyHandlerFn(cliCtx context.CLIContext) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		_, page, limit, err := rest.ParseHTTPArgsWithLimit(r, 0)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		cliCtx, ok := rest.ParseQueryHeightOrReturnBadRequest(w, cliCtx, r)
+		if !ok {
+			return
+		}
+
+		params := types.NewQueryTotalSupplyParams(page, limit)
+		bz, err := cliCtx.Codec.MarshalJSON(params)
+		if err != nil {
+			rest.WriteErrorResponse(w, http.StatusBadRequest, err.Error())
+			return
+		}
+
+		res, _, err := cliCtx.QueryWithData(fmt.Sprintf("custom/%s/%s", types.QuerierRoute, types.QueryTotalSupply), bz)
```

Comment:
Related to #4573 , should be `res, height, err` and then `ctx = ctx.WithHeight(height)` on line 59 before `PostProcessResponse` call. I can update on my pr once this is merged.